proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,428 @@
+
+Consider the folowing two subfolders.
+
+Directory One:
+
+```
+reboot-server.cs
+check-server-status.cs
+```
+
+Directory Two:
+
+```
+reboot-server.cs
+utils.cs
+```
+
+In Directory Two, the `utils.cs` file is used by `reboot-server.cs`. A developer may expect to be able to just run `dotnet run` in that directory and have the program execute. But if they are in Directory One, they would have to be explicit about which file they are running.
+
+### Coding against a framework alternative syntax
+
+The walkthrough proposes the `#!` directive `#!/usr/bin/share/dotnet-aspnet`, but the .NET CLI doesn't have a command to "pull in the ASP.NET Core references", so it wouldn't map to something like that.
+
+An alternative could be to extend `#r` like so:
+
+* `#r "framework:aspnet"`
+* `#r "sdk:aspnet"`
+* etc.
+
+This could cleanly separate the "tell the compiler which references to pull in" world from the "short-cut having to use `dotnet`" world like so:
+
+```csharp
+#!/usr/bin/share/dotnet
+#r "framework:aspnet"
+
+var app = WebApplication.CreateBuilder(args);
+app.MapGet("/", () => "Hello World");
+await app.RunAsync();
+```
+
+### `#!` and its semantics
+
+The `#!` directives in the walkthrough all point at `/usr/bin/share/dotnet`. This path may be different depending on your development OS (e.g., Fedora).
+
+What this means is that it identifies the .NET environment and implies a default set of references available in the file. However, it does not bring in other things like ASP.NET or other frameworkreferences. It is a contextual, "this is how I run" sort of directive.
+
+It could be extended/changed in a few ways:
+
+* `/usr/bin/share/dotnet run` makes it explicit that the command is to run the file
+* `/usr/bin/share/dotnet-aspnet` is an alternative to `#!aspnet` that brings ASP.NET Core into scope for the file (substitute a different framework/sdk here)
+
+In effect, the `#!` directive's semantics could span a spectrum of "I need the path and command to literally execute this file" towards "I add context that affects the references available to this file".
+
+We'll need to decide if they can affect the set of references available in the file or not.
+
+### Using `#!` to point to `dotnet` or point to `env`
+
+An alternative to `#!/path/to/dotnet` is `#!/usr/bin/env dotnet`. That would mean that any UNIX program could run this file:
+
+```csharp
+#!/usr/bin/env dotnet
+WriteLine("Hello, world!");
+```
+
+However, the downside is that you can't pass any more arguments to the `dotnet` process. It's a matter of portability vs. utility.
+
+### Build output
+
+The walkthrough states that the build output is a single executable file in the current working directory. This is based on the hypothesis that this is the least surprising place to put the executable.
+
+* Is this hypothesis based on reality?
+* Is this still true if you pull in a package?
```

Comment:
For a normal package based project, the build output consists of several files, only a subset of which are actually required to run the result:
- First we have the resulting DLL. (.NET Core does not create managed EXEs).
- There is (optionally, enabled by default) a pdb file with debugging information.
- There is an executable host that runs the the dll.
- There is the deps.json file that specifies dependency dlls.
- There is the runtimeconfig.json file that specifies which runtime to use to run this, and which shared frameworks are needed.
- There is a runtimeconfig.dev.json file that specifies alternative locations to look for assemblies, such as the nuget cache folders.
- Lastly (optionally, enabled by default) it copies referenced dlls from nuget packages. This is not needed, since they can be accessed from the nuget cache. It can be disabled with `<CopyLocalLockFileAssemblies>false</CopyLocalLockFileAssemblies>`, and everything still works.

This many files is quite simply too many to place along side the code.

As an alternative, publish could be used under the hood.  If you request creating a framework-dependent "single file" executable, on windows you can get as few as 2 files, the .exe and the .pdb (and of course the pdb generation could be supressed). (I've not tested, but I'm guessing Linux also results in 2 files, and I've no clue about macOS). Nuget referenced dlls get embeded in the exe, and the exe relies on the shared runtime installation. This might well be the result that makes the most sense for simple program builds.